bool blocked
bool end_of_row
float64 force
float64 carefull
